### PR TITLE
Extract plugins for yarn and ruby to separate responsibilities

### PR DESCRIPTION
**Pull request by xjunior**
This PR splits CobraCommander into three packages:

- `cobra_commander`: the core models of a cobra app (components and umbrella), the CLI, and tools such as graph, tree, and the executor
- `cobra_commander-ruby`: the source for ruby packages within a cobra app
- `cobra_commander-yarn`: the source for JS/yarn workspaces packages within a cobra app

With this change, a cobra app should now include the core library and the necessary plugins within the `cobra` bundle group.

The plugins should gain extra responsibility in the near future, like the introduction of Cobra Generators and dependency management (add/remove)

### Todo

- [x] Split cobra into multiple plugins
- [x] Write proper API documentation for core classes
- [x] Update README documentation

### CODE REVIEW

File: cobra_commander-ruby/lib/cobra_commander/ruby/version.rb [hunk 1]
```diff
@@ -0,0 +1,7 @@
+# frozen_string_literal: true
+
+module CobraCommander
+  module Ruby
+    VERSION = "0.1.0"
```
Comment (garettarrowood) [inline cobra_commander-ruby/lib/cobra_commander/ruby/version.rb:5]: Any reason for `0.1.0` versus `1.0.0`?

File: cobra_commander-ruby/sig/cobra_commander/ruby.rbs [hunk 2]
```diff
@@ -0,0 +1,6 @@
+module CobraCommander
+  module Ruby
+    VERSION: String
```
Comment (garettarrowood) [inline cobra_commander-ruby/sig/cobra_commander/ruby.rbs:3]: Yum!

Comment (garettarrowood) [reply cobra_commander-ruby/sig/cobra_commander/ruby.rbs:3]: Plan on setting types for other files?

File: cobra_commander-yarn/README.md [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+
+    $ bundle add cobra_commander-yarn
+
+If bundler is not being used to manage dependencies, install the gem by executing:
+
+    $ gem install cobra_commander-yarn
+
+## Usage
+
+TODO: Write usage instructions here
```
Comment (garettarrowood) [inline cobra_commander-yarn/README.md:19]: Maybe address the TODOs in the two READMEs? Anything brief will do.

File: cobra_commander-ruby/sig/cobra_commander/ruby.rbs (hunk 2 above)
Comment (garettarrowood) [reply cobra_commander-ruby/sig/cobra_commander/ruby.rbs:3]: As discussed - let's just remove this until we plan to support Ruby typing.

File: cobra_commander-ruby/LICENSE.txt [hunk 4]
```diff
@@ -0,0 +1,21 @@
+The MIT License (MIT)
+
+Copyright (c) 2022 Carlos Palhares
```
Comment (wadewinningham) [inline cobra_commander-ruby/LICENSE.txt:3]: Should this be Power Home Remodeling instead?

File: cobra_commander-ruby/lib/cobra_commander/ruby/version.rb (hunk 1 above)
Comment (xjunior) [reply cobra_commander-ruby/lib/cobra_commander/ruby/version.rb:5]: @garettarrowood no reason, I'll bump it up.

File: cobra_commander-yarn/README.md (hunk 3 above)
Comment (xjunior) [reply cobra_commander-yarn/README.md:19]: Forgot about the new README's, thanks.

File: cobra_commander-yarn/.standard.yml [hunk 5]
```diff
@@ -0,0 +1,3 @@
+# For available configuration options, see:
+#   https://github.com/testdouble/standard
+ruby_version: 2.6
```
[the remark this answers is not in the document]
Comment (xjunior) [reply cobra_commander-yarn/.standard.yml:3]: Standard is also not our standard :)

File: cobra_commander-ruby/spec/dummy/components/sales/sales.gemspec [hunk 6]
```diff
@@ -0,0 +1,12 @@
+# frozen_string_literal: true
+
+Gem::Specification.new do |spec|
+  spec.name = "sales"
+  spec.version = "0.1.0"
+  spec.summary = "This is an example component for cobra_commander-ruby"
+  spec.authors = ["Carlos Palhares"]
+
```
Comment (wadewinningham) [inline cobra_commander-ruby/spec/dummy/components/sales/sales.gemspec:8]: Should we specify the `spec.license` here, too, like with the others?

File: cobra_commander-ruby/README.md [hunk 7]
```diff
@@ -0,0 +1,33 @@
+# CobraCommander::Ruby
+
+Welcome to your new gem! In this directory, you'll find the files you need to be able to package up your Ruby library into a gem. Put your Ruby code in the file `lib/cobra_commander/ruby`. To experiment with that code, run `bin/console` for an interactive prompt.
+
+TODO: Delete this and the text above, and describe your gem
```
[the remark this answers is not in the document]
Comment (xjunior) [reply cobra_commander-ruby/README.md:5]: I'll do a better job on the README files of each gem. Thanks!

File: cobra_commander-yarn/.standard.yml (hunk 5 above)
[the remark this answers is not in the document]
Comment (garettarrowood) [reply cobra_commander-yarn/.standard.yml]: Makes sense, haha. Somehow I forgot about `standardrb`.

File: cobra_commander/spec/cobra_commander/cli_spec.rb [hunk 8]
```diff
@@ -154,50 +122,53 @@
...
+        ├── auth
+        ├── directory
+        │   └── auth
+        └── finance
+            ├── auth
+            ├── directory
+            │   └── auth
+            └── payroll
+                └── directory
+                    └── auth
```
Comment (garettarrowood) [inline cobra_commander/spec/cobra_commander/cli_spec.rb:152]: These file trees look a lot more realistic when are fixtures have names instead of single letters.

File: cobra_commander/lib/cobra_commander/component.rb [hunk 9]
```diff
@@ -9,18 +9,16 @@ def initialize(umbrella, name)
       @umbrella = umbrella
       @name = name
       @dependency_names = []
-      @packages = {}
+      @packages = []
```
Comment (garettarrowood) [inline cobra_commander/lib/cobra_commander/component.rb:12]: Performance tweak? Why the change from a dictionary to a collection?

Comment (garettarrowood) [reply cobra_commander/lib/cobra_commander/component.rb:12]: I think I see now... `source` == `key`.

File: cobra_commander/lib/cobra_commander/registry.rb [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+      Class.new(self) do
+        define_singleton_method(:key) { key }
+        define_method(:key) { key }
+
+        def self.inherited(base)
+          super
+          superclass.all[key] = base
+        end
+      end
+    end
```
Comment (garettarrowood) [inline cobra_commander/lib/cobra_commander/registry.rb:15]: I'm going to have to open a console in here to unpack what this accomplishes. I wasn't familiar with [Forwardable](https://ruby-doc.org/stdlib-2.5.1/libdoc/forwardable/rdoc/Forwardable.html).

What is `key`? Is it a name or a path, or what?

**Review by garettarrowood — APPROVED** [review]
Nice work!

File: cobra_commander/spec/cobra_commander/registry_spec.rb [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+
+  it "allows creating inheritance based registration to the registry" do
+    expect(Animal.all[:dog]).to be Dog
+    expect(Animal.all[:cat]).to be Cat
+  end
+
+  it "has different registries for defent classes" do
+    expect(Animal.all.keys).to eql %i[dog cat]
+    expect(People.all).to be_empty
+  end
```
Comment (garettarrowood) [inline cobra_commander/spec/cobra_commander/registry_spec.rb:40]: This file is helpful

File: cobra_commander/lib/cobra_commander/registry.rb (hunk 10 above)
Comment (xjunior) [reply cobra_commander/lib/cobra_commander/registry.rb:15]: Yeah, I'm not a big fan of how this is right now, but what it achieves is:

I want a class extending the registry to be **registered in the registry** and **have a key attribute** used to identify the registered class uniquely within the registry. This key is used to toggle the plugin on/off in the, for instance.

The key is a required class attribute of the source, so I didn't want to just have them define a key, I wanted to initialize it with the inheritance. This is the best I could do.

This defines a class with a key, and once inherited it will register the new class with the given key. The resulting syntax is similar to what AR does with migrations. I wish the code was simpler, but I like the overall solution.

Comment (garettarrowood) [reply cobra_commander/lib/cobra_commander/registry.rb:15]: Is it a `category` then perhaps? Or possibly a `klass`? --- Maybe even just a `registration_key`?
